[WIP] Channel safely (#2111)
DFHack/dfhack: docs/plugins/channel-safely.rst

channel-safely
==============

.. dfhack-tool::

>     :summary: Auto-manage channel designations to keep dwarves safe

Author: ```suggestion
    :summary: Auto-manage channel designations to keep dwarves safe.
```

channel-safely
==============

.. dfhack-tool::
    :summary: Auto-manage channel designations to keep dwarves safe
    :tags: fort auto

Multi-level channel projects can be dangerous, and managing the safety of your
dwarves throughout the completion of such projects can be difficult and time

> consuming. This plugin keeps your dwarves safe (while channeling) so you don't
> have to. Now you can focus on designing your dwarven cities with the deep chasms
> they were meant to have.

Author: ```suggestion
consuming. This plugin keeps your dwarves safe (at least while channeling) so you don't
have to. Now you can focus on designing your dwarven cities with the deep chasms
they were meant to have.
```

Usage
-----

::
    enable channel-safely
    channel-safely set <setting> <value>
    channel-safely enable|disable <feature>
    channel-safely runonce

When enabled the map will be scanned for channel designations which will be grouped
together based on adjacency and z-level. These groups will then be analyzed for safety
and designations deemed unsafe will be put into :wiki:`Marker Mode <Designations_menu#Marker_Mode>`.
Each time a channel designation is completed its group status is checked, and if the group
is complete pending groups below are made active again.

Examples
--------

``channel-safely``
    The plugin reports its configured status.

``channel-safely runonce``
    Runs the safety procedures once. You can use this if you prefer initiating scans manually.

``channel-safely disable require-vision``
    Allows the plugin to read all tiles, including the ones your dwarves know nothing about.

> ``channel-safely enable monitor-active``
>     Enables monitoring active channel digging jobs. Meaning that if another unit it present
>     or the tile below becomes open space the job will be paused or canceled (respectively).
> 
> ``channel-safely set ignore-threshold 3``
>     Configures the plugin to ignore designations equal to or above priority 3 designations.
> 
> Features
> --------
> 
> :require-vision:    Toggle whether the dwarves need vision of a tile before channeling to it can be deemed unsafe. (default: enabled)
> :monitor:           Toggle whether to monitor the conditions of active digs. (default: disabled)

Author: ```suggestion
``channel-safely enable monitor``
    Enables monitoring active channel digging jobs. Meaning that if another unit it present
    or the tile below becomes open space the job will be paused or canceled (respectively).

``channel-safely set ignore-threshold 3``
    Configures the plugin to ignore designations equal to or above priority 3 designations.

Features
--------

:require-vision:    Toggle whether the dwarves need vision of a tile before channeling to it can be deemed unsafe. (default: enabled)
:monitor:           Toggle whether to monitor the conditions of active digs. (default: disabled)
```

:resurrect:         Toggle whether to resurrect units involved in cave-ins, and if monitor is enabled
                    units who die while digging. (default: disabled)
:insta-dig:         Toggle whether to use insta-digging on unreachable designations.
                    Runs on the refresh cycles. (default: disabled)

Settings
--------

:refresh-freq:      The rate at which full refreshes are performed.
                    This can be expensive if you're undertaking many mega projects. (default:600, twice a day)
:monitor-freq:      The rate at which active jobs are monitored. (default:1)
:ignore-threshold:  Sets the priority threshold below which designations are processed. You can set to 1 or 0 to

>                     effectively disable the scanning. (default: 5)

Author: "below which designations are processed"

>                     effectively disable the scanning. (default: 5)

Reviewer: Oi ok human parsing error disregard

channel-safely
==============

.. dfhack-tool::
    :summary: Auto-manage channel designations to keep dwarves safe.
    :tags: fort auto

Multi-level channel projects can be dangerous, and managing the safety of your
dwarves throughout the completion of such projects can be difficult and time
consuming. This plugin keeps your dwarves safe (at least while channeling) so you don't
have to. Now you can focus on designing your dwarven cities with the deep chasms
they were meant to have.

Usage
-----

::

>     enable channel-safely

Reviewer: needs blank line here. current text has parsing errors:
![image](https://user-images.githubusercontent.com/977482/203421328-cff0a54a-b7f0-4ba2-8c7a-6eed55f94b89.png)

                    the marker mode isn't being set correctly (mostly for debugging).

Features
--------

:require-vision:    Toggle whether the dwarves need vision of a tile before channeling to it can be deemed unsafe. (default: enabled)
:monitor:           Toggle whether to monitor the conditions of active digs. (default: disabled)
:resurrect:         Toggle whether to resurrect units involved in cave-ins, and if monitor is enabled
                    units who die while digging. (default: disabled)
:insta-dig:         Toggle whether to use insta-digging on unreachable designations.
                    Runs on the refresh cycles. (default: disabled)

Settings
--------

:refresh-freq:      The rate at which full refreshes are performed.
                    This can be expensive if you're undertaking many mega projects. (default:600, twice a day)
:monitor-freq:      The rate at which active jobs are monitored. (default:1)
:ignore-threshold:  Sets the priority threshold below which designations are processed. You can set to 1 or 0 to
                    effectively disable the scanning. (default: 5)

> :fall-threshold:    Sets the fall threshold beyond which is considered unsafe. (default: 1)

Author: They all are saved yes. I thought it was mentioned, I'll update the doc

Each time a channel designation is completed its group status is checked, and if the group
is complete pending groups below are made active again.

> Settings will persist according to the save game being used.

Reviewer: I find this wording a bit hard to follow. Autolabor says this: `Once you have enabled it in a fortress, it stays enabled until you explicitly disable it, even if you save and reload your game.` which is more verbose, but also more clear.

Author: I find that wording misleading. The plugin doesn't stay enabled, just the features.

Author: I could:
```
Features and settings once set will persist until you change them, even if you save and reload your game.
```
Does that work for you?

Reviewer: that's good. and plugin enablement will be persisted soon (tm) once I get that working for all plugins.

Author: does it have an off switch?

Author: ```suggestion
Features and settings once set will persist until you change them, even if you save and reload your game.
```
